src/Storages/StorageURL.cpp

```diff
@@ -255,24 +262,94 @@ namespace
 
                 setCredentials(credentials, request_uri);
 
+                const auto settings = context->getSettings();
                 try
                 {
+                    if (download_threads > 1)
+                    {
+                        try
+                        {
+                            ReadWriteBufferFromHTTP buffer(
+                                request_uri,
+                                Poco::Net::HTTPRequest::HTTP_HEAD,
+                                callback,
+                                timeouts,
+                                credentials,
+                                settings.max_http_get_redirects,
+                                DBMS_DEFAULT_BUFFER_SIZE,
+                                read_settings,
+                                headers,
+                                ReadWriteBufferFromHTTP::Range{},
+                                context->getRemoteHostFilter(),
+                                true,
+                                /* use_external_buffer */ false,
+                                /* skip_url_not_found_error */ skip_url_not_found_error);
+
+                            Poco::Net::HTTPResponse res;
+                            buffer.call(res, Poco::Net::HTTPRequest::HTTP_HEAD);
```

Comment:
Currently `call` method from http read buffer throws exception only for retryable errors, for non-retryable errors it just sets `initialization_error` type and sets `current_exception`. So looks like need to adjust it, if `call` is called outside of http read buffer scope.
Also may be it makes sense to check for retryable errors and retry call().